docs/cpp/source/notes/tensor_cuda_stream.rst

```diff
@@ -0,0 +1,189 @@
+Tensor CUDA Stream API 
+======================
+
+A `CUDA Stream`_ is a linear sequence of execution that belongs to a specific device.
+Pytorch C++ API has defined a CUDAStream class and a few useful functions to support CUDA Stream operations.
+You can find them in `CUDAStream.h`_. This note provides more details on how to use Pytorch C++ CUDA Stream APIs.
+
+.. _CUDA Stream: https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#streams
+.. _CUDAStream.h: https://pytorch.org/cppdocs/api/file_c10_cuda_CUDAStream.h.html#file-c10-cuda-cudastream-h
+.. _CUDAStreamGuard.h: https://pytorch.org/cppdocs/api/structc10_1_1cuda_1_1_c_u_d_a_stream_guard.html
+
+Get CUDA stream
+*****************
+
+Pytorch C++ API provides the following ways to get CUDA stream:
+
+1. Get a new stream from the CUDA stream pool, streams are preallocated from the pool and returned in a round-robin fashion.
+
+.. code-block:: cpp
+  
+  CUDAStream getStreamFromPool(const bool isHighPriority = false, DeviceIndex device = -1);
+
+.. tip::
+
+  You can request a stream from the high priority pool by setting isHighPriority to true, or a stream for a specific device
+  by setting device (defaulting to the current CUDA stream).
+
+2. Get the default CUDA stream, for the passed CUDA device, or for the current device if no device index is passed.
+
+.. code-block:: cpp
+  
+  CUDAStream getDefaultCUDAStream(DeviceIndex device_index = -1);
+
+.. tip::
+  
+  The default stream is where most computation occurs when you aren't explicitly using streams.
+
+3. Get the current CUDA stream, for the passed CUDA device, or for the current device if no device index is passed. 
+
+.. code-block:: cpp
+  
+  CUDAStream getCurrentCUDAStream(DeviceIndex device_index = -1);
+
+.. tip::
+  
+  The current CUDA stream will usually be the default CUDA stream for the device, but it may be different if someone
+  called ``setCurrentCUDAStream`` or used ``StreamGuard`` or ``CUDAStreamGuard``.
+
+
+
+Set CUDA stream
+*****************
+
+Pytorch C++ API provides the following ways to set CUDA stream:
+
+1. Set the current stream on the device of the passed in stream to be the passed in stream.
```

Comment:
I have some regrets about designing the API like this because the scenario of setting a CUDA stream but not switching to it is rare and this has been confusing for people. Maybe in the future we should offer switchToCUDAStream(CUDAStream stream) which sets the device to the stream's device and sets the stream on that device to the given stream?

Reply:
Yes, this is strange. I think stream guard is used to switch both device and stream. Adding switchToCUDAStream is probably more strightforward.